### DISCUSSION TITLE
Micronaut Bean Compatibility between Version 3 and 4

### DISCUSSION THREAD

**Discussion by saw303** [micronaut-projects/micronaut-core#9414] [Q&A] [answered]
I'm upgrading a Micronaut 3 application to Micronaut 4.0.0-M3, that uses external dependencies of Micronaut beans (http clients, common type converters, …) that were compiled with Micronaut 3.

When running the Micronaut 4 application I noticed that these beans are not loaded by Micronaut 4 due to a 

```
11:22:54.951 [ForkJoinPool.commonPool-worker-5] TRACE i.m.context.condition.Condition - Bean definition for type [ch.onstructive.salespersons.SalesPersonsClient$Intercepted] not loaded since it is not on the classpath
java.lang.NoClassDefFoundError: io/micronaut/inject/BeanFactory
```

Is this an expected behaviour? Is Micronaut 4 unable to load these Micronaut 3 beans, or is it  something other that I missed?

**Comment by graemerocher** [comment, marked as answer]
the Micronaut 3 module needs recompiling with Micronaut 4

**Reply by saw303** [reply]
Thank you @graemerocher for clarifying this.

And I suppose after recompiling it with Micronaut 4 it cannot be used with a Micronaut 3 application anymore, right?
